modules/dag-l1/src/main/scala/org/tessellation/dag/l1/StateChannel.scala

```diff
@@ -0,0 +1,161 @@
+    .as(OwnRoundTrigger)
+
+  private val peerBlockConsensusInputs: Stream[F, PeerBlockConsensusInput] = Stream
+    .fromQueueUnterminated(peerBlockConsensusInputQueue)
+    .evalFilter(isPeerInputValid(_))
+    .map(_.value)
+
+  private val blockConsensusInputs: Stream[F, BlockConsensusInput] =
+    ownerBlockConsensusInputs.merge(peerBlockConsensusInputs)
+
+  private val runConsensus: Pipe[F, BlockConsensusInput, FinalBlock] =
+    _.evalTap(input => logger.debug(s"Received block consensus input to process: $input"))
+      .evalMap(
+        new BlockConsensusCell[F](_, blockConsensusContext)
+          .run()
+          .handleErrorWith(e => CellError(e.getMessage).asLeft[Ω].pure[F])
+      )
+      .flatMap {
+        case Left(ce) =>
+          Stream.eval(logger.warn(ce)(s"Error occurred during some step of block consensus.")) >>
+            Stream.empty
+        case Right(ohm) =>
+          ohm match {
+            case fb @ FinalBlock(hashedBlock) =>
+              Stream
+                .eval(logger.debug(s"Block created! Hash=${hashedBlock.hash} ProofsHash=${hashedBlock.proofsHash}"))
+                .as(fb)
+            case NullTerminal => Stream.empty
+            case other =>
+              Stream.eval(logger.warn(s"Unexpected ohm in block consensus occurred: $other")) >>
+                Stream.empty
+          }
+      }
+
+  private val gossipBlock: Pipe[F, FinalBlock, FinalBlock] =
+    _.evalTap { fb =>
+      gossip.spread(fb.hashedBlock.signed).handleErrorWith(e => logger.warn(e)("Block gossip spread failed!"))
+    }
+
+  private val peerBlocks: Stream[F, FinalBlock] = Stream
+    .fromQueueUnterminated(peerBlockQueue)
+    .evalMap(_.hashWithSignatureCheck)
+    .evalTap {
+      case Left(e)  => logger.warn(e)(s"Received an invalidly signed peer block!")
+      case Right(_) => Async[F].unit
+    }
+    .collect {
+      case Right(hashedBlock) => FinalBlock(hashedBlock)
+    }
+
+  private val storeBlock: Pipe[F, FinalBlock, Unit] =
+    _.evalMap(fb => blockStorage.store(fb.hashedBlock))
+
+  private val blockAcceptance: Stream[F, Unit] = Stream
+    .awakeEvery(1.seconds)
+    .evalMap(_ => blockStorage.getWaiting)
+    .evalTap(awaiting => logger.debug(s"Pulled following blocks for acceptance ${awaiting.keySet}"))
+    .evalMap(
+      _.toList
+        .sortBy(_._2.value.height.value)
+        .traverse {
+          case (hash, signedBlock) =>
+            logger.debug(s"Acceptance of a block $hash starts!") >>
+              blockService
+                .accept(signedBlock)
+                .handleErrorWith(logger.warn(_)(s"Failed acceptance of a block with hash=$hash"))
+
+        }
+        .void
+    )
+
+  private val blockConsensus: Stream[F, Unit] =
+    blockConsensusInputs
+      .through(runConsensus)
+      .through(gossipBlock)
+      .merge(peerBlocks)
+      .through(storeBlock)
```

Comment:
I have one thought 🤔 Is there a possibility that the block will be gossiped before being stored?
We've seen issues before with blocks that should be accepted but were missing (delay in algorithm) causing unnecessary resolving

Reply:
Discussed on slack. TLDR the block storing is idempotent and whether the block comes from local source or from gossip doesn't matter.